### ISSUE TITLE
Add meeting agendas to Latest Updates

### ISSUE THREAD

**Issue by noahmanger** [18F/fec-cms#704]
So that users know what is or was on the agenda for a particular commission meeting, add all meeting agendas to the Latest Updates.

---

This issue is to scrape all meeting agendas for all meetings from http://www.fec.gov/agenda/agendas.shtml , create a new meeting agenda page model and template, and import the new data.

When scraping, we should be pulling out (when available):
- Meeting date (`mtg_date`)
- Meeting time (`mtg_time`)
- Any links to media files attached to the meeting like:
  - Video link (`full_video_url`)
  - Audio link (`full_audio`)
  - Archived captions link (`mtg_transcript`)
  - Additional meeting documents (need a new field on the model)
- Body content (need a new field on the model)

Additionally, these additional assets should be included on the model. They can be found on the index tables for each year (e.g. http://www.fec.gov/agenda/2016/agendas2016.shtml)
- Sunshine notice for the meeting (need a new field on the model)
- Draft minutes (need a new field on the model)
- Approved meeting minutes (need a new field on the model)

**Completion criteria:**
- [x] Scrape all meetings 
- [x] Replace all relative URLs in the scraped content with absolute URLs (see other scrapers for examples)
- [x] Finish the meeting agenda model and template ([started here](https://github.com/18F/fec-cms/pull/756), [mostly finished here](https://github.com/18F/fec-cms/pull/853), final template change in PR #868)
- [x] Import all agendas under `/home/updates/` as instances of the `AgendaPage` model (done as a part of the import process)

----

The work is in PR #868

**Comment by tadhg-ohiggins** [comment]
Should we include the Sunshine Act Notices (which all seem to be PDFs) here, or treat those as separate content?

**Comment by noahmanger** [comment]
We have a [separate story](https://github.com/18F/fec-cms/issues/703) for those but if it's easier to do all at once, then that works. I think we can just handle those the same way we've been handling other PDFs, by creating a page that just includes a link to the document.

**Comment by tadhg-ohiggins** [comment]
Okay, I'll treat them separately but will bear in mind that modifying the script to handle them might happen in future.

**Comment by tadhg-ohiggins** [comment]
This is a list of broken links I've found while going through the annual pages for the commission meetings.

First, the links I don't have working versions for (I'll check them off as I get working URLs):

- [ ] http://www.fec.gov/agenda/2014/approved_14-2-a.pdf
- [ ] http://www.fec.gov/agenda/2014/approved_14-1-a.pdf
- [ ] http://www.fec.gov/sunshine/2013/notice20131205.pdf
- [ ] http://www.fec.gov/sunshine/2012/notice20120711.pdf
- [ ] http://www.fec.gov/sunshine/2012/notice20120315.pdf

Second, the list of broken URLs that I'm correcting in the output of the script:

- http://www.fec.gov/agenda/2010/2011/mtgdoc1101.pdf ->
  http://www.fec.gov/agenda/2011/mtgdoc1101.pdf
- http://www.fec.gov/agenda/sunshine/2009/notice20091210.pdf ->
  http://www.fec.gov/sunshine/2009/notice20091210.pdf
- http://www.fec.gov/agenda/sunshine/2009/notice20091202.pdf ->
  http://www.fec.gov/sunshine/2009/notice20091202.pdf
- http://www.fec.gov/agenda/sunshine/2009/notice20091117.pdf ->
  http://www.fec.gov/sunshine/2009/notice20091117.pdf
- http://www.fec.gov/agenda/sunshine/2009/notice20091103.pdf ->
  http://www.fec.gov/sunshine/2009/notice20091103.pdf
- http://www.fec.gov/agenda/sunshine/2009/notice20091027.pdf ->
  http://www.fec.gov/sunshine/2009/notice20091027.pdf
- http://www.fec.gov/agenda/sunshine/2009/notice20091001.pdf ->
  http://www.fec.gov/sunshine/2009/notice20091001.pdf
- http://www.fec.gov/agenda/sunshine/2009/notice20090917.pdf ->
  http://www.fec.gov/sunshine/2009/notice20090917.pdf
- http://www.fec.gov/agenda/sunshine/2009/notice20090820.pdf ->
  http://www.fec.gov/sunshine/2009/notice20090820.pdf
- http://www.fec.gov/agenda/sunshine/2009/notice20090722.pdf ->
  http://www.fec.gov/sunshine/2009/notice20090722.pdf
- http://www.fec.gov/agenda/sunshine/2009/notice20090709.pdf ->
  http://www.fec.gov/sunshine/2009/notice20090709.pdf
- http://www.fec.gov/agenda/sunshine/2009/notice20090617.pdf ->
  http://www.fec.gov/sunshine/2009/notice20090617.pdf
- http://www.fec.gov/agenda/sunshine/2009/notice20090610.pdf ->
  http://www.fec.gov/sunshine/2009/notice20090610.pdf
- http://www.fec.gov/agenda/sunshine/2009/notice20090512.pdf ->
  http://www.fec.gov/sunshine/2009/notice20090512.pdf
- http://www.fec.gov/agenda/sunshine/2009/notice20090501.pdf ->
  http://www.fec.gov/sunshine/2009/notice20090501.pdf
- http://www.fec.gov/agenda/sunshine/2009/notice20090428.pdf ->
  http://www.fec.gov/sunshine/2009/notice20090428.pdf
- http://www.fec.gov/agenda/sunshine/2009/notice20090417.pdf ->
  http://www.fec.gov/sunshine/2009/notice20090417.pdf
- http://www.fec.gov/agenda/sunshine/2009/notice20090409.pdf ->
  http://www.fec.gov/sunshine/2009/notice20090409.pdf
- http://www.fec.gov/agenda/sunshine/2009/notice20090313.pdf ->
  http://www.fec.gov/sunshine/2009/notice20090313.pdf
- http://www.fec.gov/agenda/sunshine/2009/notice20090306.pdf ->
  http://www.fec.gov/sunshine/2009/notice20090306.pdf
- http://www.fec.gov/agenda/sunshine/2009/notice20090227.pdf ->
  http://www.fec.gov/sunshine/2009/notice20090227.pdf
- http://www.fec.gov/agenda/sunshine/2009/notice20090205.pdf ->
  http://www.fec.gov/sunshine/2009/notice20090205.pdf
- http://www.fec.gov/agenda/sunshine/2009/notice20090122.pdf ->
  http://www.fec.gov/sunshine/2009/notice20090122.pdf
- http://www.fec.gov/agenda/sunshine/2009/notice20090107.pdf ->
  http://www.fec.gov/sunshine/2009/notice20090107.pdf
- http://www.fec.gov/agenda/sunshine/2008/notice20081231.pdf ->
  http://www.fec.gov/sunshine/2008/notice20081231.pdf

**Comment by patphongs** [comment]
@tadhg-ohiggins I have fixed the broken links that you couldn't find working versions for. If you re-run your scraper, you'll see those links are fixed. 

I have also completed importing the sunshine notices and agenda documents to the content bucket. This is the pattern of how things are should map:

## Sunshine Notices
- http://www.fec.gov/sunshine/[years-2001-2013]/ -> https://s3-us-gov-west-1.amazonaws.com/cg-47928592-406c-4536-8234-99b896e8d57d/updates/sunshine-notices/[years-2001-2013]/
- http://www.fec.gov/sunshine/[years-2014-2017]/executive/ -> https://s3-us-gov-west-1.amazonaws.com/cg-47928592-406c-4536-8234-99b896e8d57d/updates/sunshine-notices/[years-2014-2017]/executive/
- http://www.fec.gov/sunshine/[years-2014-2017]/open/ -> https://s3-us-gov-west-1.amazonaws.com/cg-47928592-406c-4536-8234-99b896e8d57d/updates/sunshine-notices/[years-2014-2017]/open/

## Agenda Documents
- http://www.fec.gov/agenda/[years-2000-2013]/ -> https://s3-us-gov-west-1.amazonaws.com/cg-47928592-406c-4536-8234-99b896e8d57d/updates/agendas/[year]
- http://www.fec.gov/agenda/[years-2014-2017]/documents/ -> https://s3-us-gov-west-1.amazonaws.com/cg-47928592-406c-4536-8234-99b896e8d57d/updates/agendas/[year]
- http://www.fec.gov/agenda/[years-2015-2017]/documents/transcripts/ -> https://s3-us-gov-west-1.amazonaws.com/cg-47928592-406c-4536-8234-99b896e8d57d/updates/agendas/[year]/transcripts/

**Comment by tadhg-ohiggins** [comment]
@patphongs Thanks! I see the latter three working, but I still get 404 errors for these two:

 http://www.fec.gov/agenda/2014/approved_14-2-a.pdf
 http://www.fec.gov/agenda/2014/approved_14-1-a.pdf

**Comment by patphongs** [comment]
@tadhg-ohiggins They should be mapped here:

http://www.fec.gov/agenda/2014/approved_14-1-a.pdf -> http://www.fec.gov/agenda/2014/documents/approved_14-1-a.pdf

http://www.fec.gov/agenda/2014/approved_14-2-a.pdf -> http://www.fec.gov/agenda/2014/documents/approved_14-2-a.pdf

**Comment by johnnyporkchops** [comment]
Updated PR for agenda page model/template is here:
https://github.com/18F/fec-cms/pull/756

**Comment by tadhg-ohiggins** [comment]
I found more broken links doing the rest of the scraping; this is a list of pages with broken links and what the broken links are:
- [ ] http://www.fec.gov/agenda/2016/agenda20160816.shtml http://www.fec.gov/agenda/2016/documents/transcripts/Open_Meeting_Captions_2016_08_16.txt
- [ ] http://www.fec.gov/agenda/2013/agenda20130425.shtml http://www.fec.gov/audio/2013/2013042505.mp3
- [ ] http://www.fec.gov/agenda/2007/agenda20070222.shtml http://www.fec.gov/agenda/2007/2007/mtgdoc07-09.pdf
- [ ] http://www.fec.gov/agenda/2006/agenda20061018.shtml http://www.fec.gov/aos/2006/aor2006-27withdrawal.pdf
- [ ] http://www.fec.gov/agenda/2005/agenda20050818.shtml http://www.fec.gov/aos/2005/aor2005-10comments.pdf
- [ ] http://www.fec.gov/agenda/2004/agenda20040429.html http://www.fec.gov/aos/2004/aor2004-10.pdf
- [ ] http://www.fec.gov/agenda/2004/agenda20040429.html http://www.fec.gov/aos/2004/aor2004-08.pdf
- [ ] http://www.fec.gov/agenda/agendas2003/agenda20030814.html http://www.fec.gov/aos/2003/aor2003-15com-5.pdf

**Comment by ccostino** [comment]
@tadhg-ohiggins, the branch you were working in has been pushed remotely, right?  This way @dogweather and I can pick things up. :-)

Thanks!

**Comment by tadhg-ohiggins** [comment]
Yes, `feature/meetings-scraper` is up to date at origin.

**Comment by dogweather** [comment]
Do we have anything in the database yet for this? Or will we be first defining it with a new `Meeting` model?

**Comment by johnnyporkchops** [comment]
There has been  a model created for meeting agendas at https://github.com/18F/fec-cms/pull/756. Please let me know if you think it will work for the scraped content or any other input

**Comment by dogweather** [comment]
Thanks @johnnyporkchops I'm comparing the code with this issue's description.

**Comment by noahmanger** [comment]
Just checking in on this: is it everything making sense? Let me know if I need to reconcile anything between the work John started and the work Tadhg did.

**Comment by noahmanger** [comment]
I just spent some time with @johnnyporkchops PR reconciling the work started there with what Tadhg and I had discussed as he was starting the scraping and I updated the issue accordingly. 

A couple things to note:
- First, in addition to including the content and assets for a meeting, we should also bring in the Sunshine Notice, Draft Minutes and Approved Minutes and attach those to the meeting model. For meetings from 2004 and on, those files are referenced in [these index tables](http://www.fec.gov/agenda/2004/agendas2004.shtml). For before 2004, they're in separate tables. If that poses too much challenge to associate those earlier notices with the meeting object, we can look at other solutions for those.
- In the model work that John started, the body content is broken out into multiple blocks to compose the agenda, each of which may have text and documents associated with it. This makes sense going forward, but for scraped content I think it might be better to just go with a regular `body` field which can contain the text scraped from the agenda, for the sake of not adding complexity. What do y'all think?
- I also added a point about replacing all relative URLs in the scraped content with absolute URLs, which is what we've been doing on similar scraped content. 

How does this sound @dogweather and @johnnyporkchops ?

**Comment by tadhg-ohiggins** [comment]
Some notes:

- The scraper currently pulls in the associated Sunshine Notices, Draft Minutes, and Approved Minutes for all of the post-2004 pages.
- The scraper only takes in the full body; trying to break things up further would almost definitely require manual work, as the HTML seemed quite varied.
- The scraper currently replaces all URLs with absolute URLs. There are some known broken links (see prior comments).

**Comment by noahmanger** [comment]
Ah excellent. Thanks for clarifying!

**Comment by johnnyporkchops** [comment]
@dogweather and @noahmanger, this course of action makes sense to me. I have two questions:

1) There would be (or already is) a separate model for building  previous agendas to be used for scraping the past agendas, And another (https://github.com/18F/fec-cms/pull/756) for new agendas going forward ?
2) For the new agenda model/template:  Would this mean I need to add fields for sunshine notices and draft minutes that are not in the current model? 
I am hoping the work I did represents a step forward for building new agenda pages,  but if the whole process needs to be more efficiently re-engineered in light of  the way we are bringing in old agendas, that is, of course, fine with me.

**Comment by noahmanger** [comment]
@johnnyporkchops (1) we should be able to use the same model for both, but you need to add an optional body field, and (2) you should just need to add URL fields to that model for those docs.

**Comment by noahmanger** [comment]
Alright, merged in @johnnyporkchops 's code starting the `AgendaPage` to `develop` so that should be good to pull in to finish up the scraping/importing.

**Comment by dogweather** [comment]
I found some buggy regex-based html processing. E.g., outputting this:

```html
 <lockquote>
    <p><u>II. ADVISORY OPINION</u></p>
    blockquote
```

@ccostino Do you know what this regex is trying to do? https://github.com/18F/fec-cms/blob/develop/fec/data_loader/utils.py#L67

**Comment by ccostino** [comment]
Yeah, we've had to modify or add some regex rules here to help account for other content that wasn't quite right.  If this is producing a problem for your importer though, feel free to adjust as necessary!  This one in particular I'm not sure as I don't believe that came up with the content I had worked with previously (but the line above it definitely did, which is why I had put it in there).

FWIW, this file came from just pulling out this code from the other importers to leverage it for future ones as opposed to just copying and pasting everything repeatedly.

**Comment by ccostino** [comment]
@noahmanger I modified the description with just a couple of updates in the checklist toward the bottom.  Once the importing is done a separate task will be needed for the `update` view changes required to have these pages appear, as you mentioned before.

**Comment by noahmanger** [comment]
Sounds good. Thanks.
